AX: Implement [WebAccessibilityObjectWrapper misspellingTextMarkerRange:forward:] on iOS using new AXSearchManager functionality.

AndresGonzalezApple opened the pull request:
#### c2423ee74d49d167d457795485801bdd03bc8a33
<pre>
AX: Implement [WebAccessibilityObjectWrapper misspellingTextMarkerRange:forward:] on iOS using new AXSearchManager functionality.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278876">https://bugs.webkit.org/show_bug.cgi?id=278876</a>
&lt;<a href="https://rdar.apple.com/problem/134957919">rdar://problem/134957919</a>&gt;

Reviewed by Tyler Wilcock.

This unifies the code paths for iOS and Mac to obtain the range of a misspelling. It also improves this functionality on iOS since now misspellingTextMarkerRange:forward: is not confined to the misspellings in the current object.

As a required byproduct of this changes, we added the conversion between AXTextMarkerRanges and NSArrays of TextMarkers used on iOS to represent ranges.

* LayoutTests/accessibility/misspelling-range.html:
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::AXTextMarkerRange::platformData const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityTextMarker initWithTextMarker:cache:]):
(-[WebAccessibilityObjectWrapper stringForTextMarkers:]):
(-[WebAccessibilityObjectWrapper misspellingTextMarkerRange:forward:]):

Canonical link: <a href="https://commits.webkit.org/282959@main">https://commits.webkit.org/282959@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e41c84e02dd51f966e24a73421b39d12e546815

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64705 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44070 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17301 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68728 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15312 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66822 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15590 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52025 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10557 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67771 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40776 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55987 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32649 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37442 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13362 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14184 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59367 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13692 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70435 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8650 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13188 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59354 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8684 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56072 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59543 "Found 2 new API test failures: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme, /WebKitGTK/TestWebKitNetworkSession:/webkit/WebKitNetworkSession/proxy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14292 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7142 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/816 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39882 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40959 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42142 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40703 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
```diff
@@ -91,7 +92,40 @@
+    return @[
+        adoptNS([[WebAccessibilityTextMarker alloc] initWithTextMarker:&m_start.m_data cache:cache]).autorelease(),
+        adoptNS([[WebAccessibilityTextMarker alloc] initWithTextMarker:&m_end.m_data cache:cache]).autorelease()
+    ];
```
Does the returned array need to be autoreleased as well?

AndresGonzalezApple replied:
Yes, I should probably return a RetainPtr<NSArray> here, thanks for catching.

twilco commented on Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
```diff
@@ -91,7 +92,40 @@
+
+    RefPtr object = m_start.object();
+    ASSERT(object); // Since *this is not null.
+    auto* cache = object->axObjectCache();
```
Should we make the cache a CheckedPtr rather than a raw pointer?

AndresGonzalezApple replied:
Decided to leave it as is because the only use is to copy the pointer into WebAccessibilityTextMarker. I'm following up with another change that actually removes the pointer from WebAccessibilityTextMarker so this becomes unnecessary.

twilco replied:
Sounds good!

AndresGonzalezApple committed:
AX: Implement [WebAccessibilityObjectWrapper misspellingTextMarkerRange:forward:] on iOS using new AXSearchManager functionality.
https://bugs.webkit.org/show_bug.cgi?id=278876
<rdar://problem/134957919>

Reviewed by Tyler Wilcock.

This unifies the code paths for iOS and Mac to obtain the range of a misspelling. It also improves this functionality on iOS since now misspellingTextMarkerRange:forward: is not confined to the misspellings in the current object.

As a required byproduct of this changes, we added the conversion between AXTextMarkerRanges and NSArrays of TextMarkers used on iOS to represent ranges.

* LayoutTests/accessibility/misspelling-range.html:
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::AXTextMarkerRange::platformData const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.h:
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityTextMarker initWithTextMarker:cache:]):
(-[WebAccessibilityObjectWrapper stringForTextMarkers:]):
(-[WebAccessibilityObjectWrapper misspellingTextMarkerRange:forward:]):

Canonical link: https://commits.webkit.org/282959@main
